Create empty host tensor for MTIA device type.

egienvalue opened the pull request:
Summary: Before copying tensor from CPU memory to device memory, for MTIA device, it doesn't need to pin the host memory first.

Test Plan: See diff D48761820

Reviewed By: jackm321

Differential Revision: D48456471

fduwjj commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -761,6 +761,21 @@ void Reducer::all_reduce_local_used_map() {
         local_used_map_tmp.data_ptr() != local_used_map_.data_ptr());
     local_used_map_tmp.copy_(local_used_map_);
     local_used_map_dev_.copy_(local_used_map_tmp, true);
+  } else if (local_used_map_dev_.is_mtia()) {
+    // For MTIA device, it doesn't need to pin host memory. Here we just use
+    // normal way of creating empty tensors.
+    auto local_used_map_tmp = at::native::empty_like(
+        local_used_map_,
+        optTypeMetaToScalarType(local_used_map_.options().dtype_opt()),
+        local_used_map_.options().layout_opt(),
+        local_used_map_.options().device_opt());
+    // Paranoid asserts here because in some workloads, the pinned
+    // allocator behaves in a way we don't understand, and may be bugged.
+    // See https://github.com/pytorch/pytorch/pull/54474
+    TORCH_INTERNAL_ASSERT(
+        local_used_map_tmp.data_ptr() != local_used_map_.data_ptr());
+    local_used_map_tmp.copy_(local_used_map_);
+    local_used_map_dev_.copy_(local_used_map_tmp, true);
```
The logic itself looks good to me, if this is for testing, I think it should be ok. But if you want to merge it, I feel we might want to consolidate the logic a little bit with the cuda case.

egienvalue replied:
Yea, we plan to merge this for unblocking the training on MTIA. I try to make this part of code independent to CUDA case incase there are more changes in the future. Do you think it is better to consolidate the logic with CUDA case ?

egienvalue committed:
Create empty host tensor for MTIA device type. (#108198)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/108198

Before copying tensor from CPU memory to device memory, for MTIA device, it doesn't need to pin the host memory first.

Test Plan: See diff D48761820

Reviewed By: jackm321

Differential Revision: D48456471

fbshipit-source-id: 45a634b2bfb3067f73c87e5cbb3e6f7502fdf094

fduwjj reviewed (approved):
LGTM
